### PR TITLE
dont rewrite a block thats already in the datastore

### PR DESCRIPTION
**Pull request by whyrusleeping**
I noticed @travisperson readding files to find out their hashes again, and it was taking forever. Theres no need to write the data to disk if its already there... This fixes that!

### CODE REVIEW

File: blockservice/blockservice.go [hunk 1]
```diff
@@ -66,7 +66,15 @@ func New(bs blockstore.Blockstore, rem exchange.Interface) (*BlockService, error
...
-	err := s.Blockstore.Put(b)
+	has, err := s.Blockstore.Has(k)
+	if err != nil {
+		return "", err
+	}
+	if has {
+		return k, nil
+	}
+
+	err = s.Blockstore.Put(b)
```
Comment (jbenet) [inline blockservice/blockservice.go]: this should probably be implemented at the blockstore level.

Comment (whyrusleeping) [inline blockservice/blockservice.go]: Hrm... I guess that makes sense, so we still send out provides messages.

Comment (briantigerchow) [inline blockservice/blockservice.go]: blockstore has a write cache

https://github.com/jbenet/go-ipfs/blob/master/blocks/blockstore/write_cache.go

Comment (whyrusleeping) [inline blockservice/blockservice.go]: Oh interesting...  It is already implemented at the blockstore level. But for some reason this change made adds a lot faster if theyve already been added already... strange.

**Pull request closed without merging**